weave/integrations/langchain/langchain_test.py

```diff
@@ -0,0 +1,601 @@
+        pass
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_chain_invoke(res.calls)
+
+
+@pytest.mark.skip_clickhouse_client
+@pytest.mark.vcr(
+    filter_headers=["authorization", "x-api-key"],
+    allowed_hosts=["api.wandb.ai", "localhost", "trace.wandb.ai"],
+    before_record_request=filter_body,
+)
+@pytest.mark.asyncio
+async def test_simple_chain_astream(
+    client: WeaveClient,
+) -> None:
+    from langchain_core.prompts import PromptTemplate
+    from langchain_openai import ChatOpenAI
+
+    api_key = os.environ.get("OPENAI_API_KEY", "sk-1234567890abcdef1234567890abcdef")
+
+    llm = ChatOpenAI(
+        model_name="gpt-3.5-turbo", openai_api_key=api_key, temperature=0.0
+    )
+    prompt = PromptTemplate.from_template("1 + {number} = ")
+
+    llm_chain = prompt | llm
+    async for _ in llm_chain.astream({"number": 2}):
+        pass
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_chain_invoke(res.calls)
+
+
+def assert_correct_calls_for_chain_batch(calls: list[tsi.CallSchema]) -> None:
+    assert len(calls) == 8
+    flattened = flatten_calls(calls)
+
+    got = [(op_name_from_ref(c.op_name), d, c.parent_id) for (c, d) in flattened]
+    ids = [c.id for (c, _) in flattened]
+
+    exp = [
+        ("langchain.Chain.RunnableSequence", 0, None),
+        ("langchain.Prompt.PromptTemplate", 1, ids[0]),
+        ("langchain.Llm.ChatOpenAI", 1, ids[0]),
+        ("openai.chat.completions.create", 2, ids[2]),
+        ("langchain.Chain.RunnableSequence", 0, None),
+        ("langchain.Prompt.PromptTemplate", 1, ids[4]),
+        ("langchain.Llm.ChatOpenAI", 1, ids[4]),
+        ("openai.chat.completions.create", 2, ids[6]),
+    ]
+    assert got == exp
+
+
+@pytest.mark.skip_clickhouse_client
+@pytest.mark.vcr(
+    filter_headers=["authorization"],
+    allowed_hosts=["api.wandb.ai", "localhost", "trace.wandb.ai"],
+    before_record_request=filter_body,
+)
+def test_simple_chain_batch(
+    client: WeaveClient,
+) -> None:
+    from langchain_core.prompts import PromptTemplate
+    from langchain_openai import ChatOpenAI
+
+    api_key = os.environ.get("OPENAI_API_KEY", "sk-1234567890abcdef1234567890abcdef")
+
+    llm = ChatOpenAI(
+        model_name="gpt-3.5-turbo", openai_api_key=api_key, temperature=0.0
+    )
+    prompt = PromptTemplate.from_template("1 + {number} = ")
+
+    llm_chain = prompt | llm
+    _ = llm_chain.batch([{"number": 2}, {"number": 3}])
+
+    res = client.server.calls_query(tsi.CallsQueryReq(project_id=client._project_id()))
+    assert_correct_calls_for_chain_batch(res.calls)
+
+
+@pytest.mark.skip_clickhouse_client
+@pytest.mark.vcr(
+    filter_headers=["authorization"],
+    allowed_hosts=["api.wandb.ai", "localhost", "trace.wandb.ai"],
+    before_record_request=filter_body,
+)
+@pytest.mark.asyncio
+async def test_simple_chain_abatch(
```

Comment:
@ayulockin and @tssweeney: Again, in the async case, it fails because the Openai traces are not under the LangChain traces.

**async batch expected result:**
```
[('langchain.Chain.RunnableSequence', 0, None),
 ('langchain.Prompt.PromptTemplate', 1, 'd3d7dac1-2fe8-4d3b-9bc5-676599dfd45a'),
 ('langchain.Llm.ChatOpenAI', 1, 'd3d7dac1-2fe8-4d3b-9bc5-676599dfd45a'),
 ('openai.chat.completions.create', 2, 'ad467678-4aba-40d5-a326-d78f700c4c84'),
 ('langchain.Chain.RunnableSequence', 0, None),
 ('langchain.Prompt.PromptTemplate', 1, 'c161a098-8a74-4836-8f29-ec5c62a7abfa'),
 ('langchain.Llm.ChatOpenAI', 1, 'c161a098-8a74-4836-8f29-ec5c62a7abfa'),
 ('openai.chat.completions.create', 2, '0dde42ab-c20b-48d9-89e5-d1c816bfa3da')]
```

**async batch actual result**

```
[('langchain.Chain.RunnableSequence', 0, None),
 ('langchain.Prompt.PromptTemplate', 1, 'd3d7dac1-2fe8-4d3b-9bc5-676599dfd45a'),
 ('langchain.Llm.ChatOpenAI', 1, 'd3d7dac1-2fe8-4d3b-9bc5-676599dfd45a'),
 ('langchain.Chain.RunnableSequence', 0, None),
 ('langchain.Prompt.PromptTemplate', 1, '7a96502c-b23e-41c7-a97d-ba1c890dc344'),
 ('langchain.Llm.ChatOpenAI', 1, '7a96502c-b23e-41c7-a97d-ba1c890dc344'),
 ('openai.chat.completions.create', 0, None),
 ('openai.chat.completions.create', 0, None)]
```